server/odc-service/src/main/java/com/oceanbase/odc/service/flow/FlowInstanceService.java

```diff
@@ -657,14 +657,23 @@ private FlowInstanceDetailResp cancel(@NotNull FlowInstance flowInstance, Boolea
             }
         }
         if (CollectionUtils.isEmpty(approvalInstances) && CollectionUtils.isEmpty(taskInstances)) {
+            // throw new UnsupportedException(ErrorCodes.FinishedTaskNotTerminable, null,
+            // "The current task has been completed and cannot be terminated");
+            log.warn("The current task has been completed, flowInstance = {}", flowInstance);
+        }
+        // check state before do update
+        if (FlowStatus.CANCELLED != flowInstance.getStatus() && FlowStatus.COMPLETED != flowInstance.getStatus()) {
+            log.info("Flow status error, cancellation conditions are not met, forced cancellation, flowInstanceId={}, "
+                    + "nodes={}", id,
+                    instances.stream().map(t -> t.getId() + "," + t.getNodeType() + "," + t.getStatus())
+                            .collect(Collectors.toList()));
+            flowInstanceRepository.updateStatusById(id, FlowStatus.CANCELLED);
```

Comment:
is there exists lock protecton for `flowInstanceRepository.updateStatusById ` ?
if not , we may judge by `update flow_instance set status='target_status' where status='previous_status'`

Reply:
Currently, no protection for flowInstanceRepository.updateStatusById, the call perform cover semantics. FlowInstance's status change may have concurrent access, let this question open to see how many update contend sql we should correct .